staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -440,22 +433,44 @@ func isDeletionDup(a, b *Delta) *Delta {
 // queueActionLocked appends to the delta list for the object.
 // Caller must lock first.
 func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) error {
+	return f.queueReplaceActionLocked(actionType, false, obj)
+}
+
+// queueReplaceActionLocked appends to the delta list for the object.
+// Caller must lock first.
+func (f *DeltaFIFO) queueReplaceActionLocked(actionType DeltaType, isReplace bool, obj interface{}) error {
 	id, err := f.KeyOf(obj)
 	if err != nil {
 		return KeyError{obj, err}
 	}
 
 	// Every object comes through this code path once, so this is a good
-	// place to call the transform func.  If obj is a
-	// DeletedFinalStateUnknown tombstone, then the containted inner object
-	// will already have gone through the transformer, but we document that
-	// this can happen. In cases involving Replace(), such an object can
-	// come through multiple times.
+	// place to call the transform func.
```

Comment:
I think we can be more succinct here:

```
	// Every object comes through this code path once, so this is a good
	// place to call the transform func.
	//
	// If obj is a DeletedFinalStateUnknown tombstone or the action is a Sync,
	// then the object will already have gone through the transformer.
	//
	// If objects already present in the cache are passed to Replace(),
	// the transformer must be idempotent to avoid re-mutating them,
	// or coordinate with all readers from the cache to avoid data races.
	// Default informers do not pass existing objects to Replace().
```

a lot of that should fold into docs on TransformFunc as well ... a lot of what is said there is inaccurate now, and we should also recommend they be idempotent

Reply:
(and if we separately made our in-tree use of transformer idempotent, we'd avoid data race risks as well and that could maybe be narrowly picked to 1.29 faster)

Reply:
Done

And I will open a separate PR to make the functions idempotent.